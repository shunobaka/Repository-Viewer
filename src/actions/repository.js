/**
 * @fileoverview Defines functions that return action objects used by redux to
 *    update repository state.
 */
import { REPOSITORIES_LOADED, REPOSITORIES_FILTERED } from './types';

/**
 * Returns a redux action object used to save repositories in state based on the
 *    provided repositories.
 * @param {array} repos Repository objects to be saved in redux state
 * @return {object} Redux action object
 * @export
 */
export const repositoriesLoaded = (repos) => {
  return {
    type: REPOSITORIES_LOADED,
    payload: repos,
  };
};

/**
 * Returns a redux action object used to filter repositories in state based on the
 *    provided query.
 * @param {string} query The name query by which repositories are filtered
 * @return {object} Redux action object
 * @export
 */
export const repositoriesFiltered = (query) => {
  return {
    type: REPOSITORIES_FILTERED,
    payload: query,
  };
};
