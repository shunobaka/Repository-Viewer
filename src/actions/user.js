/**
 * @fileoverview Defines functions that return action objects used by redux to
 *    update user state.
 */
import { USERS_SEARCH_RESULT, USER_LOADED } from './types';

/**
 * Returns a redux action object used to save users and username query in state
 *    based on the provided users and query.
 * @param {array} users User objects to be saved in the redux state
 * @param {*} query The username query by which the users were retrieved
 * @returns Redux action object
 * @export
 */
export const usersLoaded = (users, query) => {
  return {
    type: USERS_SEARCH_RESULT,
    payload: {
      users,
      query,
    },
  };
};

/**
 * Returns a redux action object used to save user in state based on the provided
 *    user.
 * @param {object} user User object to be saved in redux state
 * @returns Redux action object
 * @export
 */
export const userLoaded = (user) => {
  return {
    type: USER_LOADED,
    payload: user,
  };
};
