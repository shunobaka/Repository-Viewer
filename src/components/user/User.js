import React, { Fragment, useEffect, useState } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import repositoryActionCreator from '../../actionCreators/repository';
import userActionCreator from '../../actionCreators/user';
import {
  Form,
  FormControl,
  Button,
  Container,
  Row,
  Col,
} from 'react-bootstrap';
import Repositories from './Repositories';

const User = ({
  user,
  repositories,
  actions: { getRepositoriesForUser, filterRepositories, loadUser },
  match,
}) => {
  const [filterInput, setFilterInput] = useState('');

  useEffect(() => {
    loadUser(match.params.username);
    getRepositoriesForUser(match.params.username);
  }, [loadUser, getRepositoriesForUser, match.params.username]);

  const onChange = (e) => {
    setFilterInput(e.target.value);
  };

  const onSubmit = (e) => {
    e.preventDefault();

    filterRepositories(filterInput);
  };

  return (
    <Fragment>
      {user ? (
        <Fragment>
          <h2 className="text-center green-text">
            Viewing repositories for {user.username}
          </h2>
          {user.num_repos > 0 ? (
            <Fragment>
              <h4 className=" text-center green-text">
                Click repository name to view details
              </h4>
              <br />
              <Container>
                <Row className="justify-content-md-center">
                  <Col lg="auto">
                    <h4 className="text-center green-text">
                      Filter repositories by name:
                    </h4>
                    <Form inline onSubmit={onSubmit}>
                      <FormControl
                        type="text"
                        className="mr-sm-2 ml-auto"
                        value={filterInput}
                        onChange={onChange}
                      />
                      <Button type="submit" variant="outline-success mr-auto">
                        Filter
                      </Button>
                    </Form>
                  </Col>
                </Row>
              </Container>
              <br />
              <Repositories repositories={repositories} />
            </Fragment>
          ) : (
            <h4>User has no repositories</h4>
          )}
        </Fragment>
      ) : (
        <h2>User loading</h2>
      )}
    </Fragment>
  );
};

User.propTypes = {
  user: PropTypes.shape({
    username: PropTypes.string.isRequired,
    num_repos: PropTypes.number.isRequired,
  }),
  repositories: PropTypes.array.isRequired,
  actions: PropTypes.shape({
    getRepositoriesForUser: PropTypes.func.isRequired,
    filterRepositories: PropTypes.func.isRequired,
    loadUser: PropTypes.func.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      username: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};

const mapStateToProps = (state) => ({
  user: state.user.user,
  repositories: state.repository.displayed_repositories,
});

const mapDispatchToProps = (dispatch) => {
  return {
    actions: bindActionCreators(
      { ...repositoryActionCreator, ...userActionCreator },
      dispatch
    ),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(User);
