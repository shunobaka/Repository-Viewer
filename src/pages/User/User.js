/**
 * @fileoverview Defines a User react component that uses the user and
 *    repository redux states to display a page with filtering option and
 *    list of user repositories.
 */
import React, { Fragment, useEffect, useState } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
  Form,
  FormControl,
  Button,
  Container,
  Row,
  Col,
  Spinner,
} from 'react-bootstrap';

import repositoryActionCreator from '../../actionCreators/repository';
import userActionCreator from '../../actionCreators/user';
import alertActionCreator from '../../actionCreators/alert';
import Repositories from '../../components/user/Repositories';

/**
 * User react component that displays a page with filtering option and list of
 *    user repositories. Gets dynamically updated when user redux state changes.
 * @param {object} props Contains mapped user redux state, action creator functions
 *    and match object
 */
const User = ({
  user,
  repositories,
  loading,
  actions: {
    getRepositoriesForUser,
    filterRepositories,
    loadUser,
    removeAlerts,
  },
  match,
}) => {
  const [filterInput, setFilterInput] = useState('');

  useEffect(() => {
    removeAlerts();
    loadUser(match.params.username);
    getRepositoriesForUser(match.params.username);
  }, [removeAlerts, loadUser, getRepositoriesForUser, match.params.username]);

  const onChange = (e) => {
    setFilterInput(e.target.value);
  };

  const onSubmit = (e) => {
    e.preventDefault();

    filterRepositories(filterInput);
  };

  return (
    <Fragment>
      {user ? (
        <Fragment>
          <h2 className="text-center green-text">
            Viewing repositories for {user.username}
          </h2>
          {user.num_repos > 0 ? (
            <Fragment>
              <h4 className=" text-center green-text">
                Click repository name to view details
              </h4>
              <br />
              <Container>
                <Row className="justify-content-md-center">
                  <Col lg="auto">
                    <h4 className="text-center green-text">
                      Filter repositories by name:
                    </h4>
                    <Form inline onSubmit={onSubmit}>
                      <FormControl
                        type="text"
                        className="mr-sm-2 ml-auto"
                        value={filterInput}
                        onChange={onChange}
                      />
                      <Button type="submit" variant="outline-success mr-auto">
                        Filter
                      </Button>
                    </Form>
                  </Col>
                </Row>
              </Container>
              <br />
              {!loading ? (
                <Repositories repositories={repositories} />
              ) : (
                <h2 className="text-center">
                  <Spinner animation="border" variant="success" />
                </h2>
              )}
            </Fragment>
          ) : (
            <Container className="text-center">
              <br />
              <div className="alert alert-warning">
                User has no repositories!
              </div>
            </Container>
          )}
        </Fragment>
      ) : (
        <h2 className="text-center">
          <Spinner animation="border" variant="success" />
        </h2>
      )}
    </Fragment>
  );
};

User.propTypes = {
  /** The redux state user whose repositories are being displayed */
  user: PropTypes.shape({
    username: PropTypes.string.isRequired,
    num_repos: PropTypes.number.isRequired,
  }),
  /** The redux state repositories to be displayed */
  repositories: PropTypes.array.isRequired,
  /** Redux state boolean that signals loading of data */
  loading: PropTypes.bool.isRequired,
  /** Actions used to retrieve and filter repositories, load user and remove alerts */
  actions: PropTypes.shape({
    getRepositoriesForUser: PropTypes.func.isRequired,
    filterRepositories: PropTypes.func.isRequired,
    loadUser: PropTypes.func.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      username: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};

/**
 * Maps the user and repository redux states to the props consumed by the
 *    react component.
 * @param {object} state The complete redux state object
 */
const mapStateToProps = (state) => ({
  user: state.user.user,
  repositories: state.repository.displayed_repositories,
  loading: state.repository.loading,
});

/**
 * Binds the dispatch function to the action creators and maps them to the
 *    props consumed by the react component
 * @param {Function} dispatch The dispatch function
 */
const mapDispatchToProps = (dispatch) => {
  return {
    actions: bindActionCreators(
      {
        ...repositoryActionCreator,
        ...userActionCreator,
        ...alertActionCreator,
      },
      dispatch
    ),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(User);
