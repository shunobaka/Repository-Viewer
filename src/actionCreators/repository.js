import githubApi from '../utils/githubApi';
import {
  repositoriesLoaded,
  repositoriesFiltered,
} from '../actions/repository';
import { addAlert } from '../actions/alert';

const getRepositoriesForUser = (username) => async (dispatch) => {
  try {
    const res = await githubApi.get(`/users/${username}/repos`);
    const payload = res.data.map((item) => ({
      id: item.id,
      name: item.name,
      url: item.html_url,
      description: item.description,
      created_at: item.created_at,
    }));

    dispatch(repositoriesLoaded(payload));
  } catch (err) {
    dispatch(addAlert(err.msg));
  }
};

const filterRepositories = (query) => (dispatch) => {
  dispatch(repositoriesFiltered(query));
};

export default {
  getRepositoriesForUser,
  filterRepositories,
};
