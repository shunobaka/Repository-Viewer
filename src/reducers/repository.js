/**
 * @fileoverview Defines a reducer used to process redux actions related to repository state.
 */
import { REPOSITORIES_LOADED, REPOSITORIES_FILTERED } from '../actions/types';

/**
 * The initial repository redux state.
 */
const initialState = {
  repositories: [],
  displayed_repositories: [],
  loading: true,
};

/**
 * Returns a new repository state based on current state and provided action.
 * @param {array} state The redux repository state
 * @param {object} action Redux action object
 * @return {array} Updated repository state object
 * @export
 */
export default function (state = initialState, action) {
  const { type, payload } = action;

  switch (type) {
    case REPOSITORIES_LOADED:
      return {
        repositories: payload,
        displayed_repositories: payload,
        loading: false,
      };
    case REPOSITORIES_FILTERED:
      return {
        ...state,
        displayed_repositories: state.repositories.filter((item) =>
          item.name.toLowerCase().includes(payload.toLowerCase())
        ),
      };
    default:
      return state;
  }
}
