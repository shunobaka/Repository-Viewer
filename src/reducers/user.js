import { USER_LOADED, USERS_SEARCH_RESULT } from '../actions/types';

const initialState = {
  user: null,
  users: [],
  query: null,
  loading: true,
};

export default function (state = initialState, action) {
  const { type, payload } = action;

  switch (type) {
    case USER_LOADED:
      return {
        ...state,
        user: payload,
      };
    case USERS_SEARCH_RESULT:
      return {
        user: null,
        users: payload.users,
        query: payload.query,
        loading: false,
      };
    default:
      return state;
  }
}
