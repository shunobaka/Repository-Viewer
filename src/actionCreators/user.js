import githubApi from '../utils/githubApi';
import { usersLoaded, userLoaded } from '../actions/user';
import { setAlert } from '../actionCreators/alert';

const getUsers = (nameQuery) => async (dispatch) => {
  try {
    const res = await githubApi.get(`/search/users?q=${nameQuery}`);

    const payload = res.data.items.map((item) => ({
      username: item.login,
      avatar: item.avatar_url,
      id: item.id,
      num_repos: item.public_repos,
    }));

    dispatch(usersLoaded(payload, nameQuery));
  } catch (err) {
    dispatch(
      setAlert(
        'There was a problem searching for users, please try again.',
        false
      )
    );
  }
};

const loadUser = (username) => async (dispatch) => {
  try {
    const res = await githubApi.get(`/users/${username}`);

    const user = {
      id: res.data.id,
      username: res.data.login,
      avatar: res.data.avatar_url,
      num_repos: res.data.public_repos,
    };

    dispatch(userLoaded(user));
  } catch (err) {
    dispatch(setAlert('Could not retrieve user, please try again.', false));
  }
};

export default {
  getUsers,
  loadUser,
};
