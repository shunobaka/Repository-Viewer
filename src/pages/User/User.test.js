import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer, { act } from 'react-test-renderer';
import { Provider } from 'react-redux';
import User from './User';
import store from '../../store';
import { userLoaded } from '../../actions/user';
import { repositoriesLoaded } from '../../actions/repository';
import userActionCreators from '../../actionCreators/user';
import repositoryActionCreators from '../../actionCreators/repository';

/** Mock functions to prevent creating http requests */
jest.mock('../../actionCreators/repository.js');
jest.mock('../../actionCreators/user.js');

userActionCreators.loadUser.mockImplementation(() => {
  return { type: 'mock_action' };
});

repositoryActionCreators.getRepositoriesForUser.mockImplementation(() => {
  return { type: 'mock_action' };
});

describe('user snapshots', () => {
  const username = 'shunobaka';
  const mockRepos = [
    {
      id: 0,
      name: 'Test 1',
      url: 'TestUrl1',
      description: 'Description 1',
    },
    {
      id: 1,
      name: 'Test 2',
      url: 'TestUrl2',
    },
  ];
  const mockUserRepos = {
    username: username,
    id: 1,
    avatar: username,
    num_repos: 1,
  };
  const mockUserNoRepos = {
    username: username,
    id: 1,
    avatar: username,
    num_repos: 0,
  };
  const match = {
    params: {
      username,
    },
  };

  it('no user renders correctly', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <User match={match} />
          </Provider>
        </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('no repository user renders correctly', () => {
    act(() => {
      store.dispatch(userLoaded(mockUserNoRepos));
    });

    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <User match={match} />
          </Provider>
        </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('single repository user renders correctly', () => {
    act(() => {
      store.dispatch(userLoaded(mockUserRepos));
      store.dispatch(repositoriesLoaded([mockRepos[0]]));
    });

    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <User match={match} />
          </Provider>
        </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('multiple repositories user renders correctly', () => {
    act(() => {
      store.dispatch(userLoaded(mockUserRepos));
      store.dispatch(repositoriesLoaded(mockRepos));
    });

    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <User match={match} />
          </Provider>
        </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
