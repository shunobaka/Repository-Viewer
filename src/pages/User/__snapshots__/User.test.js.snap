// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user snapshots multiple repositories user renders correctly 1`] = `
Array [
  <h2
    className="text-center green-text"
  >
    Viewing repositories for 
    shunobaka
  </h2>,
  <h4
    className=" text-center green-text"
  >
    Click repository name to view details
  </h4>,
  <br />,
  <div
    className="container"
  >
    <div
      className="justify-content-md-center row"
    >
      <div
        className="col-lg-auto"
      >
        <h4
          className="text-center green-text"
        >
          Filter repositories by name:
        </h4>
        <form
          className="form-inline"
          onSubmit={[Function]}
        >
          <input
            className="mr-sm-2 ml-auto form-control"
            onChange={[Function]}
            type="text"
            value=""
          />
          <button
            className="btn btn-outline-success mr-auto"
            disabled={false}
            type="submit"
          >
            Filter
          </button>
        </form>
      </div>
    </div>
  </div>,
  <br />,
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-lg-6 offset-lg-3"
      >
        <div
          className="accordion"
        >
          <div
            className="card"
          >
            <div
              className="card-header"
              onClick={[Function]}
            >
              1.  
               
              Test 1
            </div>
            <div
              aria-expanded={null}
              className="collapse"
            >
              <div
                className="card-body"
              >
                <a
                  href="TestUrl1"
                >
                  Go to repository
                </a>
                <br />
                <br />
                <h5>
                  Description:
                </h5>
                <span>
                  Description 1
                </span>
                <br />
                <br />
                <h5>
                  Created on:
                </h5>
                <span>
                  Invalid Date
                </span>
              </div>
            </div>
          </div>
          <div
            className="card"
          >
            <div
              className="card-header"
              onClick={[Function]}
            >
              2.  
               
              Test 2
            </div>
            <div
              aria-expanded={null}
              className="collapse"
            >
              <div
                className="card-body"
              >
                <a
                  href="TestUrl2"
                >
                  Go to repository
                </a>
                <br />
                <br />
                <h5>
                  Description:
                </h5>
                <span
                  className="text-warning"
                >
                  This repository does not have a description
                </span>
                <br />
                <br />
                <h5>
                  Created on:
                </h5>
                <span>
                  Invalid Date
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`user snapshots no repository user renders correctly 1`] = `
Array [
  <h2
    className="text-center green-text"
  >
    Viewing repositories for 
    shunobaka
  </h2>,
  <div
    className="text-center container"
  >
    <br />
    <div
      className="alert alert-warning"
    >
      User has no repositories!
    </div>
  </div>,
]
`;

exports[`user snapshots no user renders correctly 1`] = `
<h2
  className="text-center"
>
  <div
    className="spinner-border text-success"
  />
</h2>
`;

exports[`user snapshots single repository user renders correctly 1`] = `
Array [
  <h2
    className="text-center green-text"
  >
    Viewing repositories for 
    shunobaka
  </h2>,
  <h4
    className=" text-center green-text"
  >
    Click repository name to view details
  </h4>,
  <br />,
  <div
    className="container"
  >
    <div
      className="justify-content-md-center row"
    >
      <div
        className="col-lg-auto"
      >
        <h4
          className="text-center green-text"
        >
          Filter repositories by name:
        </h4>
        <form
          className="form-inline"
          onSubmit={[Function]}
        >
          <input
            className="mr-sm-2 ml-auto form-control"
            onChange={[Function]}
            type="text"
            value=""
          />
          <button
            className="btn btn-outline-success mr-auto"
            disabled={false}
            type="submit"
          >
            Filter
          </button>
        </form>
      </div>
    </div>
  </div>,
  <br />,
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-lg-6 offset-lg-3"
      >
        <div
          className="accordion"
        >
          <div
            className="card"
          >
            <div
              className="card-header"
              onClick={[Function]}
            >
              1.  
               
              Test 1
            </div>
            <div
              aria-expanded={null}
              className="collapse"
            >
              <div
                className="card-body"
              >
                <a
                  href="TestUrl1"
                >
                  Go to repository
                </a>
                <br />
                <br />
                <h5>
                  Description:
                </h5>
                <span>
                  Description 1
                </span>
                <br />
                <br />
                <h5>
                  Created on:
                </h5>
                <span>
                  Invalid Date
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
